Aggressively use actual function parameters in `php_verror`

NattyNarwhal opened the pull request:
PHP errors used to not show parameter info consistently. Make it so that it uses a backtrace to get function info, similar to how exceptions work.

This makes the docref error functions' parameter argument mostly vestigal, being used only if allocation fails basically. The parameter argument may be useful in the case it is more verbose than the actual function args (is there a case?).

This is an INI option, so that this behaviour can be turned off. We do so for tests, as to avoid rewriting most EXPECTF sections. This can be changed, of course.

See GH-12048. (Updated in 2026-03-04.)

NattyNarwhal committed:
Aggressively use actual func params in verror

PHP errors used to not show parameter info consistently.
Make it so that it uses a backtrace to get function info, similar to how
exceptions work.

This makes the docref error functions' parameter argument mostly
vestigal, being used only if allocation fails basically.

Several tests will fail from the fact we include function params.
One annoyance is that _build_trace_args truncates strings according to
exception_string_param_max_len.

See GH-12048

Co-authored-by: Tim Düsterhus <tim@bastelstu.be>

Girgias commented on Zend/zend_exceptions.c:
```diff
@@ -587,6 +587,39 @@ static void _build_trace_string(smart_str *str, HashTable *ht, uint32_t num) /*
+			if (last_len != ZSTR_LEN(str.s)) {
+				ZSTR_LEN(str.s) -= 2; /* remove last ', ' */
+			}
+		} else {
+			smart_str_appends(&str, "<<invalid argument array>>");
+		}
```
Can this ever happen?

NattyNarwhal replied:
I copied it from `build_trace_string`, which emits an error, which I didn't want to do considering we'd be in this code path because of an error. It's only used there because it seems userland code can muck around and make it be defensive (`Zend/tests/bug63762.phpt`).

(The common code path could probably be shared though...)

Girgias replied:
It might make sense to extract the common part indeed, maybe put it in exception.c and corresponding header. Would allow to get rid of the newly introduced dependency on built-in functions header.

Girgias reviewed (commented):
Thanks for looking into this!

NattyNarwhal commented:
The big awful annoyance is going to be rewriting pretty much every test file. I know there's been other wide sweeping commits that changed a lot of stuff before though. Were they doing anything automated to clean those up?

iluuu1994 commented:
@NattyNarwhal `run-tests.php --bless`

NattyNarwhal commented:
That works, although paths are kinda annoying - all the absolute paths are there, and truncated, i.e. `Warning: symlink('/Users/calvin/s...', '../bad/./symlin...'):`. At least it fixed the resource IDs. I'm considering changing bless to make this easier to deal with.

Girgias commented:
> That works, although paths are kinda annoying - all the absolute paths are there, and truncated, i.e. `Warning: symlink('/Users/calvin/s...', '../bad/./symlin...'):`. At least it fixed the resource IDs. I'm considering changing bless to make this easier to deal with.

Changing bless probably makes sense as part of this PR yes, ideally it should replace those with ``%s``

NattyNarwhal commented:
So my change to bless as below:

```diff
diff --git a/scripts/dev/bless_tests.php b/scripts/dev/bless_tests.php
index fa49647fcf..2422512880 100755
--- a/scripts/dev/bless_tests.php
+++ b/scripts/dev/bless_tests.php
@@ -72,6 +72,8 @@ function normalizeOutput(string $out): string {
         'Resource ID#%d used as offset, casting to integer (%d)',
         $out);
     $out = preg_replace('/string\(\d+\) "([^"]*%d)/', 'string(%d) "$1', $out);
+    // Replace absolute paths, particularly those truncated; they're likely to have your homedir in it
+    $out = preg_replace("/'\\/.*\.\\.\\.'/", "'%s'", $out);
     $out = str_replace("\0", '%0', $out);
     return $out;
 }
```

...does work, like so:

```diff
-Warning: fileperms(): stat failed for /no/such/file/dir in %s on line %d
+Warning: fileperms('%s'): stat failed for /no/such/file/dir in %s on line %d
```

Of course, bless is also a little insensitive, so you do have to manually postprocess these (unless there's a better way to do it?)

```diff
-Warning: chmod(): %s in %s on line %d
+Warning: chmod('/etc/passwd', 511): Operation not permitted in %s on line %d
```

To speak nothing of the tests I can't run because i.e. Windows.

Girgias commented:
One thing that I just thought about is that if the parameter is going to be displayed it should be suppressed if the SensitiveParam attribute is used.

NattyNarwhal commented:
> One thing that I just thought about is that if the parameter is going to be displayed it should be suppressed if the SensitiveParam attribute is used.

The code that prints the arguments in errors is shared with code that prints backtraces i.e. on exceptions. Both cases are handled, so you'll get `Warning: odbc_connect('bogusdsn', 'user', Object(SensitiveParameterValue)): SQL error: [unixODBC][Driver Manager]Data source name not found and no default driver specified, SQL state IM002 in SQLConnect` and `Stack trace:
#0 /Users/calvin/src/chmod.php(10): sensitive(Object(SensitiveParameterValue))`.

NattyNarwhal commented:
One thing that came to mind was turning this into an INI option, and it could be off for just the test suite or by default in general if the new output is intrusive. That said, I don't think configurability is a good idea (in terms of making the option used, and the PHP stance on introducing new options in general), but bringing it up anyways.

Girgias commented:
Yeah not a fan about the INI setting :D

NattyNarwhal commented:
I've pushed changes to `ext/standard/tests/file`, mostly because it would likely have a ton of absolute paths in the new parameters, and a small enough set to check bless output without getting overwhelmed. Unfortunately, I still had to change a lot of the test outputs back (or further manually) in cases where bless was overzealous, ignorant (still a lot of absolute paths in error messages), or a bit naive (i.e. fscanf test putting format strings like `%d` in paramters, which confuses EXPECTF). The question is if there's a better way to scale this up to all the other tests.

NattyNarwhal committed:
Gate new error func arg display behind display_error_function_args

This is a useful feature, but enabling it by default requires rewriting
every PHPT file's output section. Since that would be a hellish diff to
make and to review, I think the best option is unfortunately, another
INI option. We can enable this for prod/dev recommended INIs, but make
sure it's disabled for the test runner.

This takes some inspiration from the discussion in GH-17056, which has
similar problems to this PR.

NattyNarwhal commented:
I've rebased this onto master and added an INI option to gate this behind. I don't like adding INI options, but it might be a lesser evil than touching every PHPT file.

bwoebi commented:
I don't think we should do _this_. Like, adding configs, just because it avoids updating some tests. What we were talking about is displaying the stacktrace, which anyway already has an ini.

bukka commented:
I thought about this and the INI actually makes sense here. The reason is that this is more an operational thing. It's really just not about avoiding the tests updates. The thing is that for some legacy projects where warnings happening quite often (I saw quite a few such projects in past), this can lead not only to a significant increase of the log space but to the potential compliance issues. An example of that might be a function receiving email addresses and compliance with GDPR or similar. I realise that there is such potential with stack traces already but those are usually less common than warnings in the logs.  So for some users it might be convenient to disable logging of parameters so having such INI seems reasonable to me.

NattyNarwhal commented:
I'm not sure if you're talking about GH-17056, this PR, or both. Sorry for any confusion by mentioning that PR on a commit here; the discussion in that PR was motivating me to revive this PR I had.

bukka commented:
I was talking about this PR but it could apply to both. But this one can especially increase the log size as fatal errors are not that common. However if you have lots regular warning in logs and they suddenly get all params logged, then it can increase the size and cost. So having an option to disable is a good think IMHO.

DanielEScherzer commented:
@NattyNarwhal happy to review this once rebased

DanielEScherzer reviewed (commented):
please
* add some tests to show the new behavior when this is enabled
* update the example INI files `php.ini-development` and `php.ini-production`

NattyNarwhal committed:
Add test for display_error_function_args

If this is not enabled by default for tests (like the fatal error
backtrace RFC), then at least test for it.

TimWolla commented on php.ini-development:
```diff
@@ -611,6 +611,15 @@ ignore_repeated_source = Off
+; Default Value: Off
+; Development Value: On
+; Production Value: On
+display_error_function_args = On
```
Related to Daniel's comment: Please avoid further INIs where the builtin default differs from the recommended “production default”.

TimWolla replied:
Relatedly, we should probably finally drop `zend.exception_ignore_args=Off` from `php.ini-production` /cc @bwoebi.

chschneider replied:
Do you mean `zend.exception_ignore_args=On`? Because that's what currently is in `php.ini-production` if I'm not mistaken. Do you think the information disclosure is no longer an issue?

NattyNarwhal committed:
Rename option to error_ignore_args

Per feedback from Tim on the RFC. Also rationalize the default vs.
recommended INI settings.

This does invert the semantics for the option; the if is changed
accordingly.

NattyNarwhal committed:
Fix this test since it calls PHP itself

Almost certainly a better way to do this...

NattyNarwhal committed:
Quick hack to placate Windows OpenSSL tests w/ new INI opt

NattyNarwhal committed:
Rename to error_include_args

Avoid a negative which is harder to reason about; matches RFC change

NattyNarwhal committed:
Default error_include_args to Off

RFC is going towards adding this, but disabling it by default.

NattyNarwhal commented:
RFC was approved; INI subvote indicates it should be set to 0 in the default INIs and in tests.

TimWolla commented on scripts/dev/bless_tests.php:
Can you extract this change as a separate PR?

NattyNarwhal replied:
Extracted into GH-22067, I'll rebase this PR to remove this (and squish the other fixes)

TimWolla commented on php.ini-production:
```diff
@@ -613,6 +613,15 @@ ignore_repeated_source = Off
+; additional information after the problem function's name.
+; Default Value: Off
+; Development Value: Off
+; Production Value: Off
```
Given these are all the same, they should be removed.

TimWolla commented on php.ini-development:
```diff
@@ -611,6 +611,15 @@ ignore_repeated_source = Off
+; additional information after the problem function's name.
+; Default Value: Off
+; Development Value: Off
+; Production Value: Off
```
ditto

TimWolla commented on Zend/tests/display_error_function_args.phpt:
```diff
@@ -0,0 +1,28 @@
+unlink('/');
+// Something with sensitive parameters that exists in a minimal build,
+// and also doesn't set anything in the docref call
+// XXX: May be better to provide a new zend_test func?
```
```suggestion
```

It's okay. `password_hash()` is unlikely to change. And if it ever does, we can still add a dedicated testing function.

TimWolla commented on Zend/tests/display_error_function_args.phpt:
```diff
@@ -0,0 +1,28 @@
+// Something with sensitive parameters that exists in a minimal build,
+// and also doesn't set anything in the docref call
+// XXX: May be better to provide a new zend_test func?
+password_hash("test", PASSWORD_BCRYPT, array("salt" => "123456789012345678901" . chr(0)));
```
Possible decrease the BCrypt cost to the accepted minimum (I think 4) to make the test faster.

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -588,30 +613,49 @@ static void _build_trace_string(smart_str *str, const HashTable *ht, uint32_t nu
+ZEND_API zend_string *zend_trace_function_args_to_string(const HashTable *frame) {
+	zval *tmp;
+	smart_str str = {0};
+	/* init because ASan will complain */
```
Complain about what?

NattyNarwhal replied:
Been a while, but pretty sure it complained if the smart_str never had anything initialized, which is fairly obvious.

TimWolla replied:
`smart_str str = {0};` will initialize the entire struct to zero. If it would not, appending an empty string would also be unsafe, because it works on garbage data.

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -588,30 +613,49 @@ static void _build_trace_string(smart_str *str, const HashTable *ht, uint32_t nu
+	}
+
+	smart_str_0(&str);
+	return str.s ? str.s : ZSTR_EMPTY_ALLOC();
```
```suggestion
	return smart_str_extract(&str);
```

NattyNarwhal replied:
Good idea; I cribbed the pattern from other functions in this file, so probably should file a PR to fix those up too...

TimWolla replied:
Cleaning up the other functions would be nice indeed!

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -588,30 +613,49 @@ static void _build_trace_string(smart_str *str, const HashTable *ht, uint32_t nu
 
+/* {{{ Gets the function arguments printed as a string from a backtrace frame. */
+ZEND_API zend_string *zend_trace_function_args_to_string(const HashTable *frame) {
+	zval *tmp;
```
Move down the declaration.

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -588,30 +613,49 @@ static void _build_trace_string(smart_str *str, const HashTable *ht, uint32_t nu
+ZEND_API zend_string *zend_trace_current_function_args_string(void) {
+	zend_string *dynamic_params = NULL;
+	/* get a backtrace to snarf function args */
+	zval backtrace, *first_frame;
```
Move first_frame down.

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -588,30 +613,49 @@ static void _build_trace_string(smart_str *str, const HashTable *ht, uint32_t nu
+		dynamic_params = zend_trace_function_args_to_string(Z_ARRVAL_P(first_frame));
+	}
+	zval_ptr_dtor(&backtrace);
+	/* free the string after we use it */
```
I don't understand the comment. It seems misplaced.

NattyNarwhal replied:
I think this was a victim of a refactor in an earlier version and never got cleaned up.

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -588,30 +613,45 @@ static void _build_trace_string(smart_str *str, const HashTable *ht, uint32_t nu
+	zval *first_frame = zend_hash_index_find(Z_ARRVAL(backtrace), 0);
+	if (first_frame) {
+		dynamic_params = zend_trace_function_args_to_string(Z_ARRVAL_P(first_frame));
+	}
```
With the `return NULL;` for “not an array” you are leaking memory if a non-array refcounted value is returned (for whatever reason). I suggest:
```suggestion
	if (Z_TYPE(backtrace) == IS_ARRAY) {
		zval *first_frame = zend_hash_index_find(Z_ARRVAL(backtrace), 0);
		if (first_frame) {
			dynamic_params = zend_trace_function_args_to_string(Z_ARRVAL_P(first_frame));
		}
	}
```

Since `dynamic_params` is initialized to `NULL` this should just work.

NattyNarwhal replied:
Applied.

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -588,30 +613,45 @@ static void _build_trace_string(smart_str *str, const HashTable *ht, uint32_t nu
+/* {{{ Gets the function arguments printed as a string from a backtrace frame. */
+ZEND_API zend_string *zend_trace_function_args_to_string(const HashTable *frame) {
+	smart_str str = {0};
+	smart_str_appends(&str, "");
```
This does nothing (useful):
```suggestion
```

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -542,6 +542,31 @@ static void _build_trace_args(zval *arg, smart_str *str) /* {{{ */
 }
 /* }}} */
 
+static void _build_trace_args_list(zval *tmp, smart_str *str) /* {{{ */
```
```suggestion
static void build_trace_args_list(zval *tmp, smart_str *str) /* {{{ */
```
Identifiers starting with an underscore are reserved. The `static` is sufficient to make the function private.

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -542,6 +542,31 @@ static void _build_trace_args(zval *arg, smart_str *str) /* {{{ */
+static void _build_trace_args_list(zval *tmp, smart_str *str) /* {{{ */
+{
+	if (EXPECTED(Z_TYPE_P(tmp) == IS_ARRAY)) {
+		size_t last_len = ZSTR_LEN(str->s);
```
I know you just moved the code, but this should be smart_str_get_len() (which safely handles the case where no allocation has been made yet (which might be the reason you got the ASAN issues?).

A simpler solution might also be to append the `, ` within the loop and just have a boolean value “first” or “not first”. This avoids manually touching the length of the backing string of the `smart_str`.

NattyNarwhal replied:
I've changed the loop and also made the condition an early return.

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -542,7 +539,33 @@ static void _build_trace_args(zval *arg, smart_str *str) /* {{{ */
-static void _build_trace_string(smart_str *str, const HashTable *ht, uint32_t num) /* {{{ */
+static void build_trace_args_list(zval *tmp, smart_str *str) /* {{{ */
+{
+	if (EXPECTED(Z_TYPE_P(tmp) != IS_ARRAY)) {
```
```suggestion
	if (UNEXPECTED(Z_TYPE_P(tmp) != IS_ARRAY)) {
```

TimWolla commented on Zend/zend_exceptions.c:
```diff
@@ -542,7 +539,33 @@ static void _build_trace_args(zval *arg, smart_str *str) /* {{{ */
+		if (++index != zend_hash_num_elements(ht)) {
+			smart_str_appends(str, ", ");
+		}
+	} ZEND_HASH_FOREACH_END();
```
Recounting for every iteration is not great. The following should work (but is untested):

```suggestion
	bool first = true;

	ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(tmp), zend_string *name, zval *arg) {
		if (!first) {
			smart_str_appends(str, ", ");
		}
		first = false;

		if (name) {
			smart_str_append(str, name);
			smart_str_appends(str, ": ");
		}
		build_trace_args(arg, str);
	} ZEND_HASH_FOREACH_END();
```

TimWolla reviewed (changes requested):
Thanks. We're almost there. You missed switching EXPECTED to UNEXPECTED when inverting the condition and the loop can be simplified.

DanielEScherzer commented on Zend/zend_exceptions.c:
```diff
@@ -588,27 +608,40 @@ static void _build_trace_string(smart_str *str, const HashTable *ht, uint32_t nu
+	/* get a backtrace to snarf function args */
+	zval backtrace;
+	zend_fetch_debug_backtrace(&backtrace, /* skip_last */ 0, /* options */ 0, /* limit */ 1);
+	/* can fail esp if low memory condition */
```
which call is failing? Is this that the zval type might not be an array on failure? Can you add a test case with memory exhaustion?

NattyNarwhal replied:
Yes, I think backtrace may not return an array if under memory exhaustion. I'm not sure how best to test memory exhaustion cases though.

DanielEScherzer replied:
not sure either, nevermind then

DanielEScherzer commented on php.ini-development:
```diff
@@ -611,6 +611,12 @@ ignore_repeated_source = Off
+; function upon an error. If this is off (or there was an error fetching the
+; arguments), the function providing the error may optionally provide some
+; additional information after the problem function's name.
+;error_include_args = Off
```
I'd recommend that the default be `On` for development, even if off for production

NattyNarwhal replied:
In retrospect, I should have done separate production vs. development INI value votes for the RFC. I'm not sure if I can do this after the vote?

DanielEScherzer replied:
Good point - let's go with Off for now in development too, and then if you want to send a separate PR to change the *development* suggested default to On we can see if there are any objections on the mailing list

jrfnl commented:
Looks like this is missing entries in UPGRADING and NEWS....
